File: services/player/internal/domain/showcase.go
package domain

import (
	"encoding/json"
	"time"
	"unicode/utf8"

	"github.com/ILITA-hub/animeenigma/libs/errors"
)

// ProfileShowcase is the Steam-style customizable profile "wall". One row
// per user. Blocks holds an ordered JSON array of Block (jsonb on Postgres,
// TEXT on SQLite in tests). The backend is a pure config store: content
// (posters, characters, cards, stats) is resolved on the frontend.
type ProfileShowcase struct {
	UserID    string    `gorm:"type:uuid;primaryKey" json:"user_id"`
	Blocks    string    `gorm:"type:jsonb;not null;default:'[]'" json:"-"`
	UpdatedAt time.Time `gorm:"not null;default:now()" json:"updated_at"`
}

func (ProfileShowcase) TableName() string { return "profile_showcases" }

// Block is one showcase block. Config is type-specific raw JSON, validated
// structurally by ValidateBlocks. Variant selects the display layout for the
// block; an empty string means "use the default variant".
type Block struct {
	Type    string          `json:"type"`
	Variant string          `json:"variant,omitempty"`
	Order   int             `json:"order"`
	Width   int             `json:"w,omitempty"`
	Height  int             `json:"h,omitempty"`
	Config  json.RawMessage `json:"config"`
}

const (
	BlockAbout             = "about"
	BlockFavoriteAnime     = "favorite_anime"
	BlockStats             = "stats"
	BlockFavoriteCharacter = "favorite_character"
	BlockCardCollection    = "card_collection"
	BlockContinueWatching  = "continue_watching"
	BlockOpEd              = "op_ed"
	BlockAnimeDNA          = "anime_dna"
	BlockCompatibility     = "compatibility"

	MaxBlocks     = 12
	MaxBlockItems = 12
	MaxAboutTitle = 64
	MaxAboutText  = 2000
)

// VariantAllowlist maps each block type to its permitted variants. The first
// entry is the default (used when Block.Variant is empty). Keep in sync with
// frontend src/types/showcase.ts.
var VariantAllowlist = map[string][]string{
	BlockAbout:             {"quote", "bio", "terminal", "minimal", "vn"},
	BlockFavoriteAnime:     {"row", "podium", "grid", "list", "banner"},
	BlockFavoriteCharacter: {"circles", "portraits", "hero", "hex"},
	BlockCardCollection:    {"row", "fan", "grid", "hero", "tilt3d"},
	BlockStats:             {"tiles", "rings", "bars", "strip"},
	BlockContinueWatching:  {"cards"},
	BlockOpEd:              {"grid"},
	BlockAnimeDNA:          {"bars"},
	BlockCompatibility:     {"ring"},
}

// SizeBound is the grid-cell size range for one (block type, variant).
type SizeBound struct{ MinW, MaxW, MinH, MaxH, DefW, DefH int }

// VariantSizeAllowlist maps block type → variant → size bounds (grid cells,
// W 1..4, H 1..3). Keep in sync with frontend src/types/showcase.ts VARIANT_SIZE.
var VariantSizeAllowlist = map[string]map[string]SizeBound{
	BlockAbout: {
		"quote": {2, 4, 1, 2, 2, 1}, "bio": {2, 4, 1, 2, 2, 2}, "terminal": {2, 4, 1, 2, 2, 2},
		"minimal": {2, 4, 1, 2, 2, 1}, "vn": {2, 2, 1, 1, 2, 1},
	},
	BlockFavoriteAnime: {
		"row": {2, 4, 1, 1, 4, 1}, "podium": {2, 2, 2, 2, 2, 2}, "grid": {2, 4, 1, 3, 4, 2},
		"list": {2, 2, 1, 3, 2, 2}, "banner": {2, 4, 1, 3, 2, 2},
	},
	BlockFavoriteCharacter: {
		"circles": {1, 4, 1, 3, 2, 1}, "portraits": {2, 4, 1, 3, 2, 2},
		"hero": {2, 4, 1, 3, 2, 2}, "hex": {1, 4, 1, 3, 2, 2},
	},
	BlockCardCollection: {
		"row": {2, 4, 1, 1, 2, 1}, "fan": {2, 4, 2, 3, 2, 2}, "grid": {2, 4, 1, 3, 2, 2},
		"hero": {2, 4, 1, 2, 2, 2}, "tilt3d": {2, 4, 2, 3, 3, 2},
	},
	BlockStats: {
		"tiles": {2, 2, 1, 1, 2, 1}, "rings": {2, 2, 1, 1, 2, 1},
		"bars": {2, 2, 1, 1, 2, 1}, "strip": {2, 2, 1, 1, 2, 1},
	},
	BlockContinueWatching: {"cards": {2, 4, 1, 3, 2, 2}},
	BlockOpEd:             {"grid": {2, 4, 1, 3, 2, 2}},
	BlockAnimeDNA:         {"bars": {1, 2, 1, 3, 1, 2}},
	BlockCompatibility:    {"ring": {1, 2, 1, 1, 2, 1}},
}

// SizeFor returns the bound for (type, variant); empty/unknown variant falls
// back to the type's default variant (VariantAllowlist[type][0]).
func SizeFor(blockType, variant string) SizeBound {
	byVariant := VariantSizeAllowlist[blockType]
	if b, ok := byVariant[variant]; ok {
		return b
	}
	if defaults := VariantAllowlist[blockType]; len(defaults) > 0 {
		return byVariant[defaults[0]]
	}
	return SizeBound{1, 4, 1, 3, 2, 1}
}

func clampInt(v, lo, hi int) int {
	if v < lo {
		return lo
	}
	if v > hi {
		return hi
	}
	return v
}

// clampBlockSize backfills absent w/h to the variant default and clamps any
// provided value into the variant's range. Never rejects.
func clampBlockSize(b *Block) {
	sb := SizeFor(b.Type, b.Variant)
	if b.Width == 0 {
		b.Width = sb.DefW
	}
	if b.Height == 0 {
		b.Height = sb.DefH
	}
	b.Width = clampInt(b.Width, sb.MinW, sb.MaxW)
	b.Height = clampInt(b.Height, sb.MinH, sb.MaxH)
}

type aboutConfig struct {
	Title string `json:"title"`
	Text  string `json:"text"`
}
type idListConfig struct {
	AnimeIDs     []string `json:"anime_ids"`
	CardIDs      []string `json:"card_ids"`
	CharacterIDs []int    `json:"character_ids"`
}
type opEdConfig struct {
	ThemeIDs []string `json:"theme_ids"`
}

func variantAllowed(blockType, variant string) bool {
	if variant == "" {
		return true // empty ⇒ default
	}
	for _, v := range VariantAllowlist[blockType] {
		if v == variant {
			return true
		}
	}
	return false
}

// ValidateBlocks enforces structural limits only (no ownership/visibility,
// no cross-service calls — see plan Global Constraints).
func ValidateBlocks(blocks []Block) error {
	if len(blocks) > MaxBlocks {
		return errors.InvalidInput("too many showcase blocks")
	}
	for i := range blocks {
		b := &blocks[i]
		if _, known := VariantAllowlist[b.Type]; !known {
			return errors.InvalidInput("unknown showcase block type")
		}
		if !variantAllowed(b.Type, b.Variant) {
			return errors.InvalidInput("unknown variant for block type")
		}
		clampBlockSize(b)
		switch b.Type {
		case BlockStats, BlockContinueWatching, BlockAnimeDNA, BlockCompatibility:
			// auto blocks: no config required
		case BlockAbout:
			var c aboutConfig
			if err := json.Unmarshal(b.Config, &c); err != nil {
				return errors.InvalidInput("invalid about config")
			}
			if utf8.RuneCountInString(c.Title) > MaxAboutTitle {
				return errors.InvalidInput("about title too long")
			}
			if utf8.RuneCountInString(c.Text) > MaxAboutText {
				return errors.InvalidInput("about text too long")
			}
		case BlockFavoriteAnime:
			var c idListConfig
			if err := json.Unmarshal(b.Config, &c); err != nil {
				return errors.InvalidInput("invalid block config")
			}
			if len(c.AnimeIDs) > MaxBlockItems {
				return errors.InvalidInput("too many items in showcase block")
			}
		case BlockCardCollection:
			var c idListConfig
			if err := json.Unmarshal(b.Config, &c); err != nil {
				return errors.InvalidInput("invalid block config")
			}
			if len(c.CardIDs) > MaxBlockItems {
				return errors.InvalidInput("too many items in showcase block")
			}
		case BlockFavoriteCharacter:
			var c idListConfig
			if err := json.Unmarshal(b.Config, &c); err != nil {
				return errors.InvalidInput("invalid block config")
			}
			if len(c.CharacterIDs) > MaxBlockItems {
				return errors.InvalidInput("too many items in showcase block")
			}
		case BlockOpEd:
			var c opEdConfig
			if err := json.Unmarshal(b.Config, &c); err != nil {
				return errors.InvalidInput("invalid op_ed config")
			}
			if len(c.ThemeIDs) > MaxBlockItems {
				return errors.InvalidInput("too many op_ed themes")
			}
		}
	}
	return nil
}

File: frontend/web/src/types/showcase.ts
export type ShowcaseState = 'none' | 'hidden' | 'visible'

/**
 * Derive the showcase visibility state from a (enabled, block count) pair.
 * Mirrors the backend mapping (auth GetShowcaseState): empty ⇒ 'none',
 * otherwise 'visible' when published / 'hidden' when not. Single FE source of
 * truth so the post-save state update can't drift from the backend rule.
 */
export function deriveShowcaseState(enabled: boolean, count: number): ShowcaseState {
  if (count === 0) return 'none'
  return enabled ? 'visible' : 'hidden'
}

export type ShowcaseBlockType =
  | 'about'
  | 'favorite_anime'
  | 'stats'
  | 'favorite_character'
  | 'card_collection'
  | 'continue_watching'
  | 'op_ed'
  | 'anime_dna'
  | 'compatibility'

export interface AboutConfig {
  title?: string
  text?: string
}
export interface FavoriteAnimeConfig {
  anime_ids: string[]
}
export interface FavoriteCharacterConfig {
  character_ids: number[]
}
export interface CardCollectionConfig {
  card_ids: string[]
}
export type StatsConfig = Record<string, never>

export interface OpEdConfig { theme_ids: string[] }
// continue_watching / anime_dna / compatibility carry no config:
export type AutoConfig = Record<string, never>

export interface ShowcaseBlock {
  type: ShowcaseBlockType
  variant?: string
  order: number
  w?: number
  h?: number
  config: AboutConfig | FavoriteAnimeConfig | FavoriteCharacterConfig
        | CardCollectionConfig | StatsConfig | OpEdConfig | AutoConfig
}

export const MAX_SHOWCASE_BLOCKS = 12
export const MAX_SHOWCASE_ITEMS = 12

// Allowlist — MUST mirror domain.VariantAllowlist (Go). First entry = default.
export const SHOWCASE_VARIANTS: Record<ShowcaseBlockType, string[]> = {
  about: ['quote', 'bio', 'terminal', 'minimal', 'vn'],
  favorite_anime: ['row', 'podium', 'grid', 'list', 'banner'],
  favorite_character: ['circles', 'portraits', 'hero', 'hex'],
  card_collection: ['row', 'fan', 'grid', 'hero', 'tilt3d'],
  stats: ['tiles', 'rings', 'bars', 'strip'],
  continue_watching: ['cards'],
  op_ed: ['grid'],
  anime_dna: ['bars'],
  compatibility: ['ring'],
}
export const defaultVariant = (t: ShowcaseBlockType) => SHOWCASE_VARIANTS[t][0]

export interface SizeBound { minW: number; maxW: number; minH: number; maxH: number; defW: number; defH: number }
const sb = (minW: number, maxW: number, minH: number, maxH: number, defW: number, defH: number): SizeBound =>
  ({ minW, maxW, minH, maxH, defW, defH })

// MUST mirror Go domain.VariantSizeAllowlist (services/player/internal/domain/showcase.go).
export const VARIANT_SIZE: Record<ShowcaseBlockType, Record<string, SizeBound>> = {
  about: { quote: sb(2,4,1,2,2,1), bio: sb(2,4,1,2,2,2), terminal: sb(2,4,1,2,2,2), minimal: sb(2,4,1,2,2,1), vn: sb(2,2,1,1,2,1) },
  favorite_anime: { row: sb(2,4,1,1,4,1), podium: sb(2,2,2,2,2,2), grid: sb(2,4,1,3,4,2), list: sb(2,2,1,3,2,2), banner: sb(2,4,1,3,2,2) },
  favorite_character: { circles: sb(1,4,1,3,2,1), portraits: sb(2,4,1,3,2,2), hero: sb(2,4,1,3,2,2), hex: sb(1,4,1,3,2,2) },
  card_collection: { row: sb(2,4,1,1,2,1), fan: sb(2,4,2,3,2,2), grid: sb(2,4,1,3,2,2), hero: sb(2,4,1,2,2,2), tilt3d: sb(2,4,2,3,3,2) },
  stats: { tiles: sb(2,2,1,1,2,1), rings: sb(2,2,1,1,2,1), bars: sb(2,2,1,1,2,1), strip: sb(2,2,1,1,2,1) },
  continue_watching: { cards: sb(2,4,1,3,2,2) },
  op_ed: { grid: sb(2,4,1,3,2,2) },
  anime_dna: { bars: sb(1,2,1,3,1,2) },
  compatibility: { ring: sb(1,2,1,1,2,1) },
}

export const sizeFor = (t: ShowcaseBlockType, variant?: string): SizeBound =>
  VARIANT_SIZE[t][variant ?? ''] ?? VARIANT_SIZE[t][SHOWCASE_VARIANTS[t][0]]

const W_CLASS: Record<number, string> = {
  1: 'col-span-1 md:col-span-1', 2: 'col-span-2 md:col-span-2',
  3: 'col-span-2 md:col-span-3', 4: 'col-span-2 md:col-span-4',
}
const H_CLASS: Record<number, string> = { 1: 'row-span-1', 2: 'row-span-2', 3: 'row-span-3' }
export const spanClasses = (w: number, h: number): string =>
  `${W_CLASS[w] ?? W_CLASS[2]} ${H_CLASS[h] ?? H_CLASS[1]}`

export const clampSize = (t: ShowcaseBlockType, variant: string | undefined, w: number, h: number) => {
  const b = sizeFor(t, variant)
  const c = (v: number, lo: number, hi: number) => Math.max(lo, Math.min(hi, v))
  return { w: c(w || b.defW, b.minW, b.maxW), h: c(h || b.defH, b.minH, b.maxH) }
}
